backend/jobs/workers/bigquery/bq_batch_worker.py

```diff
@@ -0,0 +1,154 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""CRMint's abstract worker for dealing with BigQuery in a batch mode."""
+
+import abc
+from typing import Tuple
+
+from google.api_core import page_iterator
+
+from jobs.workers.bigquery import bq_worker
+
+
+# Param name to use when passing around a BQ table name to process between
+# workers and sub-workers.
+BQ_TABLE_TO_PROCESS_PARAM = 'bq_table_to_process'
+
+# Param name to use when passing around a BQ page token between workers
+# and sub-workers
+BQ_PAGE_TOKEN_PARAM = 'bq_page_token'
+
+# Param name to use when passing around batch size between workers and
+# sub-workers.
+BQ_BATCH_SIZE_PARAM = 'bq_batch_size'
+
+
+class BQBatchDataWorker(bq_worker.BQWorker, abc.ABC):
+  """Abstract worker for batching data into manageable chunks.
+
+  This worker will start the processing of a large dataset from BQ, enqueueing
+  a sub-worker to handle each chunk of data pulled from the BQ table.  If the
+  original instance enqueues too many sub-workers, it will enqueue another
+  instance of itself and stop processing.
+
+  In order to use this batch processing, an implementer needs to do the
+  following:
+    1)  Create an implementation of the 'TablePageResultsProcessorWorker'
+        sub-worker to handle processing each page of data.
+    2)  Create an implementation of this base class with the
+        'get_sub_worker_name' function, so the base worker knows which
+        sub-worker to enqueue.
+    3)  Add the sub-worker class name to the sub-workers private workers
+        mapping.
+    4)  Add the base class to the public workers mapping.
+  """
+
+  # BigQuery batch size for querying results.
+  DEFAULT_BQ_BATCH_SIZE = 1000
+
+  # Maximum number of jobs to enqueued before spawning a new scheduler.
+  MAX_ENQUEUED_JOBS_PER_COORDINATOR = 50
+
+  def _execute(self) -> None:
+    table_name_to_process = self._generate_qualified_bq_table_name()
+    self.log_info(
+      f'About to query BQ table for conversions: {table_name_to_process}')
+    page_token = self._params.get(BQ_PAGE_TOKEN_PARAM, None)
+    client = self._get_client()
+
+    row_iterator = client.list_rows(
+      table=client.get_table(table_name_to_process),
+      page_token=page_token,
+      page_size=self.DEFAULT_BQ_BATCH_SIZE
+    )
+
+    enqueued_jobs_count = 0
+    for _ in row_iterator.pages:
+
+      # Enqueue job for this page
+      worker_params = self._params.copy()
+      worker_params[BQ_PAGE_TOKEN_PARAM] = page_token
+      worker_params[BQ_BATCH_SIZE_PARAM] = self.DEFAULT_BQ_BATCH_SIZE
```

Comment:
It seems "bq_batch_size" param is ignored for now. Perhaps we should implement this as an actual default.